swarm/pss/pss_test.go

```diff
@@ -265,6 +266,82 @@ func TestKeys(t *testing.T) {
 	//psp := ps.symKeyPool[inkeyid][topic]
 }
 
+type pssTestPeer struct {
+	*protocols.Peer
+	addr []byte
+}
+
+func (t *pssTestPeer) Address() []byte {
+	return t.addr
+}
+
+func (t *pssTestPeer) Update(addr network.OverlayAddr) network.OverlayAddr {
+	return addr
+}
+
+func (t *pssTestPeer) Off() network.OverlayAddr {
+	return &pssTestPeer{}
+}
+
+// forwarding should skip peers that do not have matching pss capabilities
+func TestMismatch(t *testing.T) {
+
+	// create privkey for forwarder node
+	privkey, err := crypto.GenerateKey()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// initialize overlay
+	baseaddr := network.RandomAddr()
+	kad := network.NewKademlia((baseaddr).Over(), network.NewKadParams())
+	rw := &p2p.MsgPipeRW{}
+
+	// one peer has a mismatching version of pss
+	wrongpssaddr := network.RandomAddr()
+	wrongpsscap := p2p.Cap{
+		Name:    pssProtocolName,
+		Version: 0,
+	}
+	nid, _ := discover.HexID("0x01")
+	wrongpsspeer := &pssTestPeer{
+		Peer: protocols.NewPeer(p2p.NewPeer(nid, common.ToHex(wrongpssaddr.Over()), []p2p.Cap{wrongpsscap}), rw, nil),
+		addr: wrongpssaddr.Over(),
+	}
+
+	// one peer doesn't even have pss (boo!)
+	nopssaddr := network.RandomAddr()
+	nopsscap := p2p.Cap{
+		Name:    "nopss",
+		Version: 1,
+	}
+	nid, _ = discover.HexID("0x02")
+	nopsspeer := &pssTestPeer{
+		Peer: protocols.NewPeer(p2p.NewPeer(nid, common.ToHex(nopssaddr.Over()), []p2p.Cap{nopsscap}), rw, nil),
+		addr: nopssaddr.Over(),
+	}
+
+	// add peers to kademlia and activate them
+	// it's safe so don't check errors
+	kad.Register([]network.OverlayAddr{wrongpsspeer})
+	kad.On(wrongpsspeer)
+	kad.Register([]network.OverlayAddr{nopsspeer})
+	kad.On(nopsspeer)
+
+	// create pss
+	pssmsg := &PssMsg{
+		To:      []byte{},
+		Expire:  uint32(time.Now().Add(time.Second).Unix()),
+		Payload: &whisper.Envelope{},
+	}
+	ps := newTestPss(privkey, kad, nil)
+
+	// run the forward
+	// it is enough that it completes; trying to send to incapable peers would create segfault
+	ps.forward(pssmsg)
```

Comment:
It may be good to check for error here. Even if only the segfault is tested, the test case is testing for capabilities, and it would help to ensure that both segfault and error are absent.

Reply:
the thing is that as it is now, `*Pss.Forward()` is not set to actually return any error. This is due to how the rest of the callers are coded (bubbling up an error from `forward` could terminate the protocol even if it's not warranted). We could return, but then I should change the corresponding caller codes in the core. Do you reckon it's that important?

Reply:
Thanks for the clarification. I see now that the check is irrelevant.